### PR TITLE
Versioning for custom action operations

### PR DESCRIPTION
**Pull request by snopoke**
## Description
Support versioning of custom actions (operations) linked to experiments. 

Quote from [README.md](https://github.com/dimagi/open-chat-studio/blob/99571a1dfcc867bff0fb35cb90b322c394d9e086/apps/custom_actions/README.md)

The `CustomAction` model is not versioned. However, the `CustomActionOperation` model is versioned since this is the
model that is linked to the Experiment.

**Working Version**

When a custom action or operation is first created, it is in a "working version" state. In this
state, the schema is not persisted to the database but is computed dynamically when needed.
This means that changes to the `CustomAction` schema will be reflected in the working version of the Experiment.

**Persisted Version**

When a new version of an experiment is created, the custom action operations are versioned. The
schema for each operation is then persisted to the database, ensuring that the exact state of the operation at the
time of versioning is preserved regardless of changes to the `CustomAction`. It also ensures that the operation is still
valid when the version is created.

### Demo
<div>
    <a href="https://www.loom.com/share/cbd7c586c8114a758e2a5921cf2b32b5">
      <p>Experiment Versioning Demo with Custom Actions 👩‍🔬 - Watch Video</p>
    </a>
    <a href="https://www.loom.com/share/cbd7c586c8114a758e2a5921cf2b32b5">
      
    </a>
  </div>

### CODE REVIEW

**Comment by snopoke** [comment]
@coderabbit review

**Comment by snopoke** [comment]
@coderabbit review

**Comment by snopoke** [comment]
@coderabbit resolve

**Review by SmittieC — DISMISSED** [review]
Nice. Looks good to me. 1 non-blocking question. Is the only reason we're not saving the `_operation_schema` for working version `CustomActionOperation` records the fact that if the associated `CustomAction` changes, we'd have to update this?

File: apps/custom_actions/models.py [hunk 1]
```diff
@@ -70,7 +74,18 @@ def get_operations_by_id(self):
         return {op.operation_id: op for op in self.operations}
 
 
-class CustomActionOperation(models.Model):
+class CustomActionOperationManager(VersionsObjectManagerMixin, models.Manager):
+    pass
+
+
+class CustomActionOperation(models.Model, VersionsMixin, VersioningMixin):
```
Comment (SmittieC) [inline apps/custom_actions/models.py:81]: Maybe `VersioningMixin` should be collapsed onto `BaseModel`? There's only 1 method in there

Comment (snopoke) [reply apps/custom_actions/models.py:81]: That wouldn't help here since this model doesn't extend `BaseModel` but maybe it should?

**Comment by snopoke** [comment]
> Is the only reason we're not saving the `_operation_schema` for working version `CustomActionOperation` records the fact that if the associated `CustomAction` changes, we'd have to update this?

Yup, that's correct.

File: apps/custom_actions/models.py (hunk 1 above)
Comment (SmittieC) [reply apps/custom_actions/models.py:81]: I don't see why not? My philosohy is to rather have those created_at / updated_at timestamps from BaseModel than not.

**Comment by snopoke** [comment]
putting this back in draft mode so that I can address the assistant versioning changes once those are merged

File: apps/utils/models.py [hunk 2]
```diff
@@ -1,8 +1,10 @@
+from typing import Self
+
 from django.db import models
 
 
 class VersioningMixin:
-    def compare_with_model(self, new: models.Model, exclude_fields: list[str]) -> set:
+    def compare_with_model(self, new: Self, exclude_fields: list[str]) -> set[str]:
```
Comment (SmittieC) [inline apps/utils/models.py:7]: `Self` as a type. The more you know

**Review by SmittieC — APPROVED** [review]
Nice

**Pull request merged**